### PR TITLE
Clean up and fix default sampling bug

### PR DESCRIPTION
**Pull request by youngbupark**
# Description

Clean up and fix default sampling bug.
- Fixed default sampling rate bug - it always set samplingrate to 0.0 before
- Renamed methods to be explcit.
- Deleted unnecessary methods.
- No tracing when samplingrate is 0.0.
- Some other clean ups and refactoring. 

## Issue reference

#1607 

## Checklist

Please make sure you've  completed the relevant tasks for this PR, out of the following list:

* [x] Code compiles correctly
* [x] Created/updated tests
* [x] Unit tests passing
* [ ] End-to-end tests passing
* [ ] Extended the documentation / Created issue in the https://github.com/dapr/docs/ repo: dapr/docs#_[issue number]_
* [ ] Specification has been updated / Created issue in the https://github.com/dapr/docs/ repo: dapr/docs#_[issue number]_
* [ ] Provided sample for the feature / Created issue in the https://github.com/dapr/docs/ repo: dapr/docs#_[issue number]_

### CODE REVIEW

**Comment by yaron2** [comment]
The linter is sad.

File: pkg/diagnostics/grpc_tracing.go [hunk 1]
```diff
@@ -26,41 +26,35 @@ const grpcTraceContextKey = "grpc-trace-bin"
...
 		if isInternalCalls(info.FullMethod) {
-			_, span = StartTracingServerSpanFromGRPCContext(newCtx, spanName, spec)
+			// For dapr.proto.internals package, this generates ServerSpan.
+			spanKind = trace.WithSpanKind(trace.SpanKindServer)
 		} else {
-			// Instead of generating the span in direct_messaging, dapr changes the spanname
-			// to CallLocal.
+			// For dapr.proto.runtime package, this generates ClientSpan because each Dapr service API
+			// is treat as dependency or internal calls.
 			if spanName == "/dapr.proto.runtime.v1.Dapr/InvokeService" {
```
Comment (yaron2) [inline pkg/diagnostics/grpc_tracing.go:45]: We're hard coding the package version `v1` here. If this changes, the runtime will break.

File: pkg/diagnostics/grpc_tracing.go [hunk 2]
```diff
@@ -127,25 +93,15 @@ func FromGRPCContext(ctx context.Context) (trace.SpanContext, bool) {
 	return sc, ok
 }
 
-// AppendToOutgoingGRPCContext appends binary serialized SpanContext to the outgoing GRPC context
-func AppendToOutgoingGRPCContext(ctx context.Context, spanContext trace.SpanContext) context.Context {
+// SpanContextToGRPCMetadata appends binary serialized SpanContext to the outgoing GRPC context
+func SpanContextToGRPCMetadata(ctx context.Context, spanContext trace.SpanContext) context.Context {
+	if (spanContext == trace.SpanContext{}) {
```
Comment (yaron2) [inline pkg/diagnostics/grpc_tracing.go:98]: What are we testing here?

File: pkg/diagnostics/http_tracing.go [hunk 3]
```diff
@@ -109,9 +90,11 @@ func SpanContextFromRequest(req *fasthttp.Request) (sc trace.SpanContext, ok boo
 
 // SpanContextToRequest modifies the given request to include traceparent and tracestate headers.
 func SpanContextToRequest(sc trace.SpanContext, req *fasthttp.Request) {
-	h := SpanContextToString(sc)
-	req.Header.Set(traceparentHeader, h)
-	tracestateToRequest(sc, req)
+	if (sc != trace.SpanContext{}) {
```
Comment (yaron2) [inline pkg/diagnostics/http_tracing.go:93]: Same as above, under which condition is this expected to be `!=`?

File: pkg/diagnostics/grpc_tracing.go (hunk 2 above)
Comment (youngbupark) [reply pkg/diagnostics/grpc_tracing.go:98]: trace is disabled and spanContext is empty. so no ops.

File: pkg/diagnostics/http_tracing.go (hunk 3 above)
Comment (youngbupark) [reply pkg/diagnostics/http_tracing.go:93]: correct. sc will be empty object and no ops if trace is disabled.

**Comment by youngbupark** [comment]
@yaron2 this refactoring and cleanup will give you 3.15 ms number with metrics-enabled.

File: pkg/diagnostics/grpc_tracing.go (hunk 1 above)
Comment (youngbupark) [reply pkg/diagnostics/grpc_tracing.go:45]: I can fix it

**Comment by youngbupark** [comment]
> @yaron2 this refactoring and cleanup will give you 3.15 ms number with metrics-enabled.

Even tracing is enabled. it gave me 3.15ms.

**Comment by yaron2** [comment]
> > @yaron2 this refactoring and cleanup will give you 3.15 ms number with metrics-enabled.
> 
> Even tracing is enabled. it gave me 3.15ms.

Sounds too good to be true :)

**Comment by youngbupark** [comment]
> > > @yaron2 this refactoring and cleanup will give you 3.15 ms number with metrics-enabled.
> > 
> > 
> > Even tracing is enabled. it gave me 3.15ms.
> 
> Sounds too good to be true :)

Check this result - https://github.com/dapr/dapr/runs/742894234?check_suite_focus=true

image

File: pkg/diagnostics/http_tracing.go (hunk 3 above)
Comment (yaron2) [reply pkg/diagnostics/http_tracing.go:108]: You probably need parentheses here for the equality check:

https://stackoverflow.com/questions/28447297/how-to-check-for-an-empty-struct

File: pkg/diagnostics/grpc_tracing.go (hunk 2 above)
Comment (yaron2) [reply pkg/diagnostics/grpc_tracing.go:124]: You probably need parentheses here for the equality check:

https://stackoverflow.com/questions/28447297/how-to-check-for-an-empty-struct

Comment (youngbupark) [reply pkg/diagnostics/grpc_tracing.go:124]: this is the way OC code uses. but I think I have the better way.

Comment (youngbupark) [reply pkg/diagnostics/grpc_tracing.go]: I used `(trace.SpanContext{}) == spanContext` which is one of option in the link you refer to.

and added unit-tests for this.

File: pkg/diagnostics/http_tracing.go (hunk 3 above)
Comment (youngbupark) [reply pkg/diagnostics/http_tracing.go]: fixed and added unit-tests

**Comment by youngbupark** [comment]
@yaron2 @shalabhms Please review it again. I added more tests and rename middlewares consistently. No logical change.

**Pull request merged**